wallet/datastore.go

```diff
@@ -0,0 +1,333 @@
+		return handlers.WrapError(err, "unable to link wallet", http.StatusBadRequest)
+	}
+	providerLinkingID := uuid.NewV5(walletClaimNamespace, tx.UserID)
+	if info.ProviderLinkingID != nil {
+		// check if the member matches the associated member
+		if !uuid.Equal(*info.ProviderLinkingID, providerLinkingID) {
+			return handlers.WrapError(errors.New("wallets do not match"), "unable to match wallets", http.StatusForbidden)
+		}
+		if anonymousAddress != nil && info.AnonymousAddress != nil && !uuid.Equal(*anonymousAddress, *info.AnonymousAddress) {
+			err := service.Datastore.SetAnonymousAddress(info.ID, anonymousAddress)
+			if err != nil {
+				return handlers.WrapError(err, "unable to set anonymous address", http.StatusInternalServerError)
+			}
+		}
+	} else {
+		err := service.Datastore.LinkWallet(info.ID, providerLinkingID, anonymousAddress)
+		if err != nil {
+			status := http.StatusInternalServerError
+			if err == ErrTooManyCardsLinked {
+				status = http.StatusConflict
+			}
+			return handlers.WrapError(err, "unable to link wallets", status)
+		}
+	}
+
+	if decimal.NewFromFloat(0).LessThan(tx.Probi) {
+		_, err := service.SubmitCommitableAnonCardTransaction(ctx, info, transaction, "", true)
+		if err != nil {
+			return handlers.WrapError(err, "unable to transfer tokens", http.StatusBadRequest)
+		}
+	}
+	return nil
+}
+
+// TxLinkWalletInfo pass a tx to set the anonymous address
+func (pg *Postgres) TxLinkWalletInfo(tx *sqlx.Tx, ID string, providerLinkingID uuid.UUID, anonymousAddress *uuid.UUID) error {
+	statement := `
+	UPDATE wallets
+	SET
+			provider_linking_id = $2,
+			anonymous_address = $3
+	WHERE id = $1;`
+	_, err := tx.Exec(
+		statement,
+		ID,
+		providerLinkingID,
+		anonymousAddress,
+	)
+	return err
+}
+
+func txGetByProviderLinkingID(tx *sqlx.Tx, providerLinkingID uuid.UUID) (*[]walletutils.Info, error) {
+	statement := `
+	SELECT *
+	FROM wallets
+	WHERE provider_linking_id = $1
+	`
+	var wallets []walletutils.Info
+	err := tx.Select(&wallets, statement, providerLinkingID)
+	return &wallets, err
+}
+
+// LinkWallet links a wallet together
+func (pg *Postgres) LinkWallet(ID string, providerLinkingID uuid.UUID, anonymousAddress *uuid.UUID) error {
+	tx, err := pg.RawDB().Beginx()
+	if err != nil {
+		return err
+	}
+	defer pg.RollbackTx(tx)
+
+	walletsMatchingProviderLinkingID, err := txGetByProviderLinkingID(tx, providerLinkingID)
+	if err != nil {
+		return errorutils.Wrap(err, "error looking up wallets by provider id")
+	}
+	walletLinkedLength := len(*walletsMatchingProviderLinkingID)
+	if walletLinkedLength >= 3 {
+		if walletLinkedLength > 3 {
+			sentry.WithScope(func(scope *sentry.Scope) {
+				anonAddr := ""
+				if anonymousAddress != nil {
+					anonAddr = anonymousAddress.String()
+				}
+				scope.SetTags(map[string]string{
+					"walletId":          ID,
+					"providerLinkingId": providerLinkingID.String(),
+					"anonymousAddress":  anonAddr,
+				})
+				sentry.CaptureMessage("too many cards linked")
```

Comment:
Is it possible that a normal user would run into this case? Would we consider this an error that we want to be alerted on? I wonder if instead of putting this in Sentry we should track this in Prometheus as a metric?

Reply:
:+1:

Reply:
https://github.com/brave-intl/bat-go/pull/314/commits/d9ad61f5ecbbf6e774a35d6481720162bacc3f87